src/Microsoft.VisualStudio.Editor.Razor/DefaultImportDocumentManager.cs

```diff
@@ -0,0 +1,158 @@
+                Debug.Assert(importFilePath != null);
+
+                if (!_importTrackerCache.TryGetValue(importFilePath, out var importTracker))
+                {
+                    // First time seeing this import. Start tracking it.
+                    var fileChangeTracker = _fileChangeTrackerFactory.Create(importFilePath);
+                    importTracker = new ImportTracker(fileChangeTracker);
+                    _importTrackerCache[importFilePath] = importTracker;
+
+                    fileChangeTracker.Changed += FileChangeTracker_Changed;
+                    fileChangeTracker.StartListening();
+                }
+
+                importTracker.AssociatedDocuments.Add(tracker.FilePath);
+            }
+        }
+
+        public override void OnUnsubscribed(VisualStudioDocumentTracker tracker)
+        {
+            if (tracker == null)
+            {
+                throw new ArgumentNullException(nameof(tracker));
+            }
+
+            _foregroundDispatcher.AssertForegroundThread();
+
+            var imports = GetImportItems(tracker);
+            foreach (var import in imports)
+            {
+                var importFilePath = import.PhysicalPath;
+                Debug.Assert(importFilePath != null);
+
+                if (_importTrackerCache.TryGetValue(importFilePath, out var importTracker))
+                {
+                    importTracker.AssociatedDocuments.Remove(tracker.FilePath);
+
+                    if (importTracker.AssociatedDocuments.Count == 0)
+                    {
+                        // There are no open documents that care about this import. We no longer need to track it.
+                        importTracker.FileChangeTracker.StopListening();
+                        _importTrackerCache.Remove(importFilePath);
+                    }
+                }
+            }
+        }
+
+        private IEnumerable<RazorProjectItem> GetImportItems(VisualStudioDocumentTracker tracker)
+        {
+            var projectDirectory = Path.GetDirectoryName(tracker.ProjectPath);
+            var templateEngine = _templateEngineFactoryService.Create(projectDirectory, _ => { });
+            var imports = templateEngine.GetImportItems(tracker.FilePath);
+
+            return imports;
+        }
+
+        private void FireImportChanged(string importPath, FileChangeKind kind)
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            var handler = Changed;
+            if (handler != null && _importTrackerCache.TryGetValue(importPath, out var importTracker))
+            {
+                var args = new ImportChangedEventArgs(importPath, kind, importTracker.AssociatedDocuments);
+                handler(this, args);
+            }
+        }
+
+        private void FileChangeTracker_Changed(object sender, FileChangeEventArgs e)
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            FireImportChanged(e.FilePath, e.Kind);
```

Comment:
Minor but you could just inline the contents of this method. It doesn't seem to be used for testing as well.

Reply:
I like to have 1 degree of separation between firing an event and listening to an event. Makes things like constructing args for the new event cleaner (don't have to pick  different name) also we've followed the pattern so far to have an `OnEventName` which is responsible for triggering a certain event. That being said, going to rename `FireImportChanged` to `OnChanged`

Reply:
Result:

```C#
private void OnChanged(ImportTracker importTracker, FileChangeKind changeKind)
{
    _foregroundDispatcher.AssertForegroundThread();

    if (Changed == null)
    {
        return;
    }

    var args = new ImportChangedEventArgs(importTracker.FilePath, changeKind, importTracker.AssociatedDocuments);
    Changed.Invoke(this, args);
}

private void FileChangeTracker_Changed(object sender, FileChangeEventArgs args)
{
    _foregroundDispatcher.AssertForegroundThread();

    if (_importTrackerCache.TryGetValue(args.FilePath, out var importTracker))
    {
        OnChanged(importTracker, args.Kind);
    }
}
```